packages/xstate-react/src/fsm.ts

```diff
@@ -56,58 +56,54 @@ export function useMachine<
         stateMachine.config,
         options ? options : (stateMachine as any)._options
       )
-    ).start()
+    )
   );
 
   const [state, setState] = useState(() => getServiceState(service));
 
-  useEffect(() => {
+  useIsomorphicLayoutEffect(() => {
     if (options) {
       (service as any)._machine._options = options;
     }
   });
 
   useEffect(() => {
     service.subscribe(setState);
+    service.start();
+
     return () => {
       service.stop();
     };
   }, []);
 
-  return [state, service.send, service];
+  return [state, service.send, service] as any;
 }
 
-export function useService<
-  TContext extends object,
-  TEvent extends EventObject = EventObject,
-  TState extends Typestate<TContext> = { value: any; context: TContext }
->(
-  service: StateMachine.Service<TContext, TEvent, TState>
-): [
-  StateMachine.State<TContext, TEvent, TState>,
-  StateMachine.Service<TContext, TEvent, TState>['send'],
-  StateMachine.Service<TContext, TEvent, TState>
-] {
-  const subscription: Subscription<
-    StateMachine.State<TContext, TEvent, TState>
-  > = useMemo(() => {
-    let currentState = getServiceState(service);
+export function useService<TService extends AnyService>(
+  service: TService
+): [StateFrom<TService>, TService['send'], TService] {
+  const getSnapshot = useCallback(() => getServiceState(service), [service]);
 
-    return {
-      getCurrentValue: () => currentState,
-      subscribe: (callback) => {
-        const { unsubscribe } = service.subscribe((state) => {
-          if (state.changed !== false) {
-            currentState = state;
-            callback();
-          }
-        });
-        return unsubscribe;
-      }
-    };
-  }, [service]);
+  const isEqual = useCallback(
```

Comment:
Can this just be declared externally?

Reply:
ah, yes - this one here, for `/fsm`, doesn't require access to the local `service` - so it can be hoisted

good 👁️  👍